### PR TITLE
UDP Reciever merge to Main

### LINKED ISSUE #11: reduce UDP to one file

### PR DESCRIPTION
**Pull request by Spenceroni-Pepperoni**
Closes #11

### CODE REVIEW

File: src/main/java/frc/robot/BallDataPacket.java [hunk 1]
```diff
@@ -0,0 +1,24 @@
+package frc.robot;
+
+import java.nio.ByteBuffer;
+
+import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
+
```
Comment (liam-a-wms) [inline src/main/java/frc/robot/BallDataPacket.java]: Add class documentation.

File: src/main/java/frc/robot/BallDataPacket.java [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+public class BallDataPacket implements DataPacket {
+    float x, y;
+    int ballDetect;
+
```
Comment (liam-a-wms) [inline src/main/java/frc/robot/BallDataPacket.java]: Add method documentation describing the expected byte formatting.

File: src/main/java/frc/robot/DataPacket.java [hunk 3]
```diff
@@ -0,0 +1,5 @@
+package frc.robot;
+
+public interface DataPacket {
```
Comment (liam-a-wms) [inline src/main/java/frc/robot/DataPacket.java]: Add class and method documentation.

File: src/main/java/frc/robot/GenericBuffer.java [hunk 4]
```diff
@@ -0,0 +1,19 @@
+package frc.robot;
+
+import java.util.ArrayList;
+
```
Comment (liam-a-wms) [inline src/main/java/frc/robot/GenericBuffer.java]: Add class documentation.

File: src/main/java/frc/robot/GenericBuffer.java [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+public class GenericBuffer<T extends DataPacket> {
+    ArrayList<DataPacket> buffer;
+
+    public GenericBuffer() {
```
Comment (liam-a-wms) [inline src/main/java/frc/robot/GenericBuffer.java]: Add constructor documentation.

File: src/main/java/frc/robot/GenericBuffer.java [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+        buffer = new ArrayList<>();
+    }
+
+    public synchronized void addData(DataPacket packet) {
```
Comment (liam-a-wms) [inline src/main/java/frc/robot/GenericBuffer.java]: Add method documentation.

File: src/main/java/frc/robot/GenericBuffer.java [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+        buffer.add(packet);
+    }
+
+    public synchronized T getData() {
```
Comment (liam-a-wms) [inline src/main/java/frc/robot/GenericBuffer.java]: Add method documentation.

File: src/main/java/frc/robot/LimelightDataPacket.java [hunk 8]
```diff
@@ -0,0 +1,22 @@
+package frc.robot;
+
+import java.nio.ByteBuffer;
+
+import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
+
+public class LimelightDataPacket implements DataPacket {
```
Comment (liam-a-wms) [inline src/main/java/frc/robot/LimelightDataPacket.java]: Add class documentation.

File: src/main/java/frc/robot/LimelightDataPacket.java [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+    double x, y;
+    int ballDetect;
+
+    @Override
+    public DataPacket fromBytes(byte[] bytes) {
```
Comment (liam-a-wms) [inline src/main/java/frc/robot/LimelightDataPacket.java]: Add method documentation describing expected data format.

File: src/main/java/frc/robot/RobotContainer.java [hunk 10]
```diff
@@ -47,11 +47,19 @@
  */
 public class RobotContainer {
   // The robot's subsystems and commands are defined here...
-  //private final UDPRecieverSubsystem m_udpsubsystem = new UDPRecieverSubsystem();
-
+  private final UDPRecieverSubsystem m_udpsubsystem = new UDPRecieverSubsystem();
```
Comment (liam-a-wms) [inline src/main/java/frc/robot/RobotContainer.java]: Referencing a deleted file. Delete this line.

File: src/main/java/frc/robot/UDPReciever.java [hunk 11]
```diff
@@ -4,21 +4,29 @@
 import java.net.DatagramPacket;
 import java.net.DatagramSocket;
 import java.net.SocketException;
+import java.util.function.Supplier;
 
-public class UDPReceiverBall extends Thread {
-   // CommsBufferBall CommunicationsBuffer = new CommsBufferBall();
+import frc.robot.DataPacket;
+
+public class UDPReciever<T extends DataPacket> extends Thread {
```
Comment (liam-a-wms) [inline src/main/java/frc/robot/UDPReciever.java]: Add class documentation.

File: src/main/java/frc/robot/UDPReciever.java [hunk 12]
```diff
@@ -4,21 +4,29 @@
...
+    final int port;
     DatagramSocket socket1;
     DatagramPacket dat1;
     byte[] receiveData1 = new byte[20];
+    final Supplier<T> dataPacketConstructor;
+    GenericBuffer buf;
+
+    public UDPReciever(int port, Supplier<T> dataPacketConstructor, GenericBuffer<T> buf) {
```
Comment (liam-a-wms) [inline src/main/java/frc/robot/UDPReciever.java]: Add constructor documentation.

File: src/main/java/frc/robot/UDPReciever.java [hunk 13]
```diff
@@ -4,21 +4,29 @@
...
-            socket1 = new DatagramSocket(5802);
+            socket1 = new DatagramSocket(port);
             dat1 = new DatagramPacket(receiveData1, receiveData1.length);
         } catch (SocketException e) {
-
         }
+
     }
 
     @Override
```
Comment (liam-a-wms) [inline src/main/java/frc/robot/UDPReciever.java]: Add method documentation.

**Review by liam-a-wms — CHANGES_REQUESTED** [review]

**Comment by Spenceroni-Pepperoni** [comment]
Added documentation comments and cleaned up code

**Review by SixtyFiveOhTwo — CHANGES_REQUESTED** [review]
Please create a folder in the robot directory, something like "UDP", and move the data packet classes, generic buffer, and UDPReciver classes into it for organization. In addition, id also like to see this run just to confirm that there is no more loop overrun errors.

**Comment by Spenceroni-Pepperoni** [comment]
I tested the code on the bunnybot yesterday and I didn't end up getting overrun errors

**Review by liam-a-wms — APPROVED** [review]

**Review by SixtyFiveOhTwo — APPROVED** [review]

**Pull request merged**